File: impl_python/parser_nodes.py
from typing import List, Optional, Any, Union
from dataclasses import dataclass

@dataclass
class Node:
  line: int
  column: int


@dataclass
class Statement(Node): pass

@dataclass
class StmtBlock(Node):
    statements: List[Statement]

@dataclass
class Expr(Node):
  tokens: List[Node] # The RPN stream

@dataclass
class Identifier(Node):
    name: Node

@dataclass
class IdentifierList(Node):
    identifiers: List[Identifier]

@dataclass
class SetStmt(Statement):
  identifiers: IdentifierList    # Variable names
  expr: Expr    # The RPN logic

@dataclass
class OutputStmt(Statement):
  expr: Expr    # The RPN logic

@dataclass
class ErrorStmt(Statement):
  expr: Expr    # The RPN logic

@dataclass
class ReturnStmt(Statement):
    expr: Expr

@dataclass
class ExitStmt(Statement):
    condition: Expr
    expr: Expr


@dataclass
class FuncStmt(Statement):
    name: str
    args: IdentifierList
    exit_stmt : ExitStmt
    body_stmt: StmtBlock
    return_stmt: ReturnStmt

@dataclass
class ImportStmt(Statement):
  filename : Identifier

@dataclass
class FunctionPointer:
    name: Identifier

@dataclass
class Integer:
    value: int

@dataclass
class Program:
    source_path: str
    #type_definitions: dict  # To store those 'type' structures
    body: StmtBlock  # The main entry point of the code

File: impl_julia/parser_nodes.jl
# 1. Define the Abstract Hierarchy Categories
abstract type ASTNode end
abstract type Statement <: ASTNode end
abstract type Expression <: ASTNode end

# 2. Base Structural Nodes
struct StmtBlock <: ASTNode
    line::Int
    column::Int
    statements::Vector{Statement}
end

struct Expr <: Expression
    line::Int
    column::Int
    tokens::Vector{ASTNode} # The RPN stream containing Identifiers, Integers, etc.
end

struct Identifier <: ASTNode
    line::Int
    column::Int
    name::String
end

struct IdentifierList <: ASTNode
    line::Int
    column::Int
    identifiers::Vector{Identifier}
end

# 3. Concrete Statements (All subtype '<:' Statement)
struct SetStmt <: Statement
    line::Int
    column::Int
    identifiers::IdentifierList
    expr::Expr
end

struct OutputStmt <: Statement
    line::Int
    column::Int
    expr::Expr
end

struct ErrorStmt <: Statement
    line::Int
    column::Int
    expr::Expr
end

struct ReturnStmt <: Statement
    line::Int
    column::Int
    expr::Expr
end

struct ExitStmt <: Statement
    line::Int
    column::Int
    condition::Expr
    expr::Expr
end

struct FuncStmt <: Statement
    line::Int
    column::Int
    name::String
    args::IdentifierList
    exit_stmt::Union{ExitStmt, Nothing} # Optional component
    body_stmt::StmtBlock
    return_stmt::ReturnStmt
end

struct ImportStmt <: Statement
    line::Int
    column::Int
    filename::Identifier
end

# 4. Primitive Types / Value Leaf Nodes
struct FunctionPointer <: ASTNode
    line::Int
    column::Int
    name::Identifier
end

struct IntegerNode <: ASTNode # Renamed to IntegerNode to avoid conflict with Julia's built-in Integer type
    line::Int
    column::Int
    value::Int64
end

struct Program <: ASTNode
    source_path::String
    body::StmtBlock
end
